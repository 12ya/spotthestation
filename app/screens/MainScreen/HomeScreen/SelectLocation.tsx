/* eslint-disable camelcase */
import React, { useEffect, useState } from "react"
import { ViewStyle, View, Pressable, PressableProps, TextStyle, ScrollView } from "react-native"

import { Accessory, Icon, Text, TextField } from "../../../components"
import { colors, typography } from "../../../theme"
import { ExpandContainer } from "../components/ExpandContainer"
import { ListItem } from "../components/ListItem"
import { useSafeAreaInsetsStyle } from "../../../utils/useSafeAreaInsetsStyle"
import { getCurrentLocation, getNearbyPlaces, getPlaces } from "../../../utils/geolocation"
import { LocationType } from "../../OnboardingScreen/SignupLocation"
import Snackbar from "react-native-snackbar"

export interface SelectLocationProps {
  /**
   * A function for closing modal.
   */
  onClose?: PressableProps["onPress"]
}

export function SelectLocation({ onClose }: SelectLocationProps) {
  const [isFocus, setIsFocus] = useState(false)
  const [textValue, setTextValue] = useState("")
  const [current, setCurrent] = useState<LocationType | null>(null)
  const [nearby, setNearby] = useState<LocationType[]>([])
  const [searchResult, setSearchResult] = useState<LocationType[]>([])

  const $marginTop = useSafeAreaInsetsStyle(["top"], "margin")
  const $paddingBottom = useSafeAreaInsetsStyle(["bottom"], "padding")

  const getLocations = async () => {
    const location = await getCurrentLocation()
    const res = await getNearbyPlaces(location.location, 100)
    setCurrent(location)
    setNearby(res)
  }

  const setPlaces = async (value: string) => {
    const location = await getPlaces(value)
    
    setSearchResult(location)
  }

  useEffect(() => {
    getLocations().catch((e: Error) => {
      Snackbar.show({
        text: e.message || "Some error occured",
        duration: Snackbar.LENGTH_SHORT,
      })
    })
  }, [])

  useEffect(() => {
    setPlaces(textValue).catch((e: Error) => {
      Snackbar.show({
        text: e.message || "Some error occured",
        duration: Snackbar.LENGTH_SHORT,
      })
    })
  }, [textValue])

  return (
    <View style={[$modalBodyContainer, $marginTop, $paddingBottom]}>
      <Pressable
        style={$close}
        onPress={onClose}
      >
        <Icon icon="x" color={colors.palette.neutral450} />
      </Pressable>
      <Text tx="homeScreen.selectLocation.title" style={$title} />
      <TextField
        value={textValue}
        autoCapitalize="none"
        autoCorrect={false}
        placeholderTx="homeScreen.selectLocation.inputPlaceholder"
        onChangeText={setTextValue}
        onFocus={() => setIsFocus(true)}
        onBlur={() => setIsFocus(false)}
        inputWrapperStyle={isFocus ? [$locations, $active] : $locations}
        renderLeftAccessory={({ style }) => (
          <Accessory 
            icon="search"
            color={colors.palette.neutral450}
            style={style}
          />
        )}
        renderRightAccessory={(({ style }) => isFocus && textValue ? (
          <Accessory 
            style={style}
            icon={"xCircle"}
            onPress={() => setTextValue("")}
          />
        ) : (
          <Accessory 
            style={style}
            icon={"currentLocation"}
          />
        ))}
      />
      <ScrollView style={$scrollContainer}>
        {(!isFocus && searchResult.length === 0) && (
          <>
            {Boolean(current) && <ExpandContainer title="homeScreen.selectLocation.current" expandble={false}>
              <ListItem icon="pin" title={current.title} subtitle={current.subtitle} selected={true} />
            </ExpandContainer>}
            <ExpandContainer title="homeScreen.selectLocation.saved" itemsCount={2}>
              <ListItem icon="pin" title="George Bush St, Times Square" subtitle="New York City, NY, 83957" />
              <ListItem icon="pin" title="Navy Garden" subtitle="San Diego, CA, 54813" />
            </ExpandContainer>
            {nearby.length > 0 && <ExpandContainer title="homeScreen.selectLocation.nearby" itemsCount={nearby.length} defaultValue={false}>
                {nearby.map((place: LocationType) => <ListItem key={place.title} icon="pin" title={place.title} subtitle={place.subtitle} />)}
              </ExpandContainer>
            }
          </>
        )}
        {(isFocus || searchResult.length > 0) && 
          <ExpandContainer title="homeScreen.selectLocation.search" itemsCount={searchResult.length} expandble={false}>
            {searchResult.map((place: LocationType) => <ListItem key={place.title} icon="pin" title={place.title} subtitle={place.subtitle} />)}
          </ExpandContainer>
        }
      </ScrollView>
    </View>
  )
}

const $modalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.neutral350,
  borderTopLeftRadius: 18,
  borderTopRightRadius: 18,
  flex: 1
}

const $scrollContainer: ViewStyle = { 
  flex: 1, 
  paddingHorizontal: 36,
  marginBottom: 24
}

const $close: ViewStyle = {
  position: "absolute",
  top: 0,
  right: 0,
  padding: 18
}

const $title: TextStyle = {
  marginTop: 36,
  marginBottom: 24,
  fontFamily: typography.primary?.normal,
  fontSize: 36,
  lineHeight: 44,
  color: colors.palette.neutral250,
  paddingHorizontal: 36,
}

const $locations: ViewStyle = {
  borderWidth: 1.5,
  borderColor: "transparent",
  borderRadius: 28,
  height: 56,
  backgroundColor: colors.palette.neutral550,
  overflow: "hidden",
  marginHorizontal: 36,
}

const $active: ViewStyle = {
  borderWidth: 1.5,
  borderColor: colors.palette.buttonBlue,
  backgroundColor: colors.palette.overlayBlue,
}
