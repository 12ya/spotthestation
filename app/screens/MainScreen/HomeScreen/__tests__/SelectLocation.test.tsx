/* eslint-disable react/display-name */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import React from "react"
import { NavigationContainer } from "@react-navigation/native"
import { SelectLocation } from "../SelectLocation"
import { render } from "@testing-library/react-native"

it("renders correctly", () => {
  const tree = render(
    <NavigationContainer>
      <SelectLocation onClose={jest.fn()} />
    </NavigationContainer>,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
