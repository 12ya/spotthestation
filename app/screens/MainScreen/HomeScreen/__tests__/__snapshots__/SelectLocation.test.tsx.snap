// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#1E2022",
        "borderTopLeftRadius": 26,
        "borderTopRightRadius": 26,
        "flex": 1,
      },
      Object {},
      Object {},
      Object {
        "fontFamily": "Helvetica Neue",
        "fontSize": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <div />
    <div />
    <div />
    <View
      style={
        Object {
          "paddingHorizontal": 16,
          "paddingTop": 10,
        }
      }
    >
      <View
        accessibilityHint="pressable settings item"
        accessibilityLabel="pressable settings item"
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingTop": 23,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "borderBottomWidth": 2,
                "borderColor": "#353739",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginLeft": 15,
                "paddingBottom": 23,
              },
              Object {
                "borderBottomWidth": 0,
              },
            ]
          }
        >
          <div />
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
                "paddingLeft": 26,
              }
            }
          >
            <div />
          </View>
          <div />
        </View>
      </View>
    </View>
    <RCTScrollView
      accessibilityHint="type for search location"
      accessibilityLabel="search location"
      accessibilityRole="scrollbar"
      accessible={true}
      style={
        Object {
          "flex": 1,
          "marginBottom": 35,
          "paddingHorizontal": 52,
        }
      }
    >
      <View />
    </RCTScrollView>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={true}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    useNativeDriverForBackdrop={true}
    visible={false}
  />
</View>
`;
